pkg/balancer/least_connections.go

```diff
@@ -0,0 +1,128 @@
+package balancer
+
+import (
+	"math"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/square/p2/pkg/util"
+)
+
+type lcBackend struct {
+	conns    int
+	disabled bool
+}
+
+type LeastConnections struct {
```

Comment:
Least connections is one of the benefits of having a centralized load balancer (client-side load balancers cannot know who is less loaded) so I wanted to highlight it. Round-robin would certainly be easier. Depending on the workloads, servers can end up better off with least-connections balancing over round-robin.

You bring up an interesting problem. The host should be removed from the pool by the monitor relatively quickly in such cases. We could potentially aggressively disable hosts where we notice that `net.Dial` is failing, but we can't detect serious issues at the HTTP level since we aren't terminating TLS.